include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h

```diff
@@ -0,0 +1,271 @@
+        backend_ = std::make_shared<Backend>();
+        initialize_impl(resample_time);
+      }
+    }
+
+    void initialize(const std::vector<resource_type>& u, double resample_time=never_resample) {
+      if (!state_) {
+        state_ = std::make_shared<state_t>();
+        backend_ = std::make_shared<Backend>(u);
+        initialize_impl(resample_time);
+      }
+    }
+
+    template<typename Function, typename ...Args>
+    selection_type select(Function&& f, Args&&...args) {
+      if (state_) {
+        std::unique_lock<std::mutex> l(state_->m_);
+        auto k =  make_task_key(std::forward<Function>(f), std::forward<Args>(args)...);
+        auto t  = state_->tuner_by_key_[k];
+        auto index = t->get_resource_to_profile();
+        if (index == use_best_resource) {
+          return selection_type{*this, t->best_resource_, t};
+        } else {
+          auto r = state_->resources_with_index_[index];
+          return selection_type{*this, r, t};
+        }
+      } else {
+         throw std::logic_error("select called before initialization");
+      }
+    }
+
+    template<typename Function, typename ...Args>
+    auto submit(selection_type e, Function&& f, Args&&... args) {
+      if (backend_) {
+        return backend_->submit(e, std::forward<Function>(f), std::forward<Args>(args)...);
+      } else {
+         throw std::logic_error("submit called before initialization");
+      }
+    }
+
+    auto get_resources() {
+       if (backend_) {
+         return backend_->get_resources();
+       } else {
+         throw std::logic_error("get_resources called before initialization");
+       }
+    }
+
+    auto get_submission_group() {
+      if (backend_) {
+        return backend_->get_submission_group();
+       } else {
+         throw std::logic_error("get_submission_group called before initialization");
+       }
+    }
+
+  private:
+
+    //
+    // types
+    //
+
+    using task_key_t = std::tuple<void *, KeyArgs...>;
+    using tuner_by_key_t = std::map<task_key_t, std::shared_ptr<tuner_t>>;
+
+    //
+    // member variables
+    //
+
+    double resample_time_ = 0.0;
+
+    struct state_t {
+      std::mutex m_;
+      std::vector<resource_with_index_t> resources_with_index_;
+      tuner_by_key_t tuner_by_key_;
+    };
+
+    std::shared_ptr<Backend> backend_;
+    std::shared_ptr<state_t> state_;
+
+    //
+    // private member functions
+    //
+
+    void initialize_impl(double resample_time=never_resample) {
+      resample_time_ = resample_time;
+      auto u = get_resources();
+      for (size_type i = 0; i < u.size(); ++i) {
+        state_->resources_with_index_.push_back(resource_with_index_t{u[i], i});
+      }
+    }
+
+    template<typename Function, typename... Args>
+    task_key_t make_task_key(Function&& f, Args&&... args) {
+      // called under lock
+      task_key_t k = std::tuple_cat(std::tuple<void *>(&f), std::make_tuple(std::forward<Args>(args)...));
```

Comment:
How about simplification?

```suggestion
      task_key_t k = std::make_tuple(std::static_cast<void*>(&f), std::forward<Args>(args)...));
```